### PR TITLE
Avx512 eltwise 

### PR DESCRIPTION
**Pull request by deeptiag1**
Resolving conflicts and Code enabled in https://github.com/libxsmm/libxsmm/pull/624
Files touched:
         samples/eltwise/run_test_avx512_vl256.sh
         src/generator_mateltwise_misc_avx_avx512.c
         src/generator_mateltwise_reduce_avx_avx512.c
         src/generator_mateltwise_sse_avx_avx512.c

### CODE REVIEW

**Comment by deeptiag1** [comment]
Checking on some seg faults

**Comment by alheinecke** [comment]
@egeor can you please have a look?

**Comment by egeor** [comment]
I am not sure why you are using both times 56, this test takes only ldi... In general, please use the test scripts for valid configs (e.g. for the opreduce kernel many configs are meaningless and the unit tests take care of such things)

**Comment by deeptiag1** [comment]
Hmm this was generated when I ran the kernel_test. However this was only generated one time.
I think then we are good, let me know if some code needs to be changed or fixed. Thanks!

**Comment by alheinecke** [comment]
how are we good, there was an issue one time and it crashes... the question is now: is it the driver or the kernel where it segfaults.

**Comment by alheinecke** [comment]
I'm confused..... I looked into this and command below is not possible on master (can you please describe what you did?):
```
[kernel_test]$ cat reduce_* | grep "eltwise_unary_reduce"
  ./eltwise_unary_reduce ${M} ${N_ADJ} ${LDI} ${REDUCE_X} ${REDUCE_X2} ${REDUCE_ROWS} ${REDUCE_OP} 1 ${N_IDX} 0
  ./eltwise_unary_reduce ${M} ${N_ADJ} 100 ${REDUCE_X} ${REDUCE_X2} ${REDUCE_ROWS} ${REDUCE_OP} 1 ${N_IDX} 0
  ./eltwise_unary_reduce ${M} ${N_ADJ} ${LDI} ${REDUCE_X} ${REDUCE_X2} ${REDUCE_ROWS} ${REDUCE_OP} 1 ${N_IDX} 0
  ./eltwise_unary_reduce ${M} ${N_ADJ} 100 ${REDUCE_X} ${REDUCE_X2} ${REDUCE_ROWS} ${REDUCE_OP} 1 ${N_IDX} 0
  ./eltwise_unary_reduce ${M} ${N_ADJ} ${LDI} ${REDUCE_X} ${REDUCE_X2} ${REDUCE_ROWS} ${REDUCE_OP} 0 ${N_IDX} 0
  ./eltwise_unary_reduce ${M} ${N_ADJ} 100 ${REDUCE_X} ${REDUCE_X2} ${REDUCE_ROWS} ${REDUCE_OP} 0 ${N_IDX} 0
  ./eltwise_unary_reduce ${M} ${N_ADJ} ${LDI} ${REDUCE_X} ${REDUCE_X2} ${REDUCE_ROWS} ${REDUCE_OP} 0 ${N_IDX} 0
  ./eltwise_unary_reduce ${M} ${N_ADJ} 100 ${REDUCE_X} ${REDUCE_X2} ${REDUCE_ROWS} ${REDUCE_OP} 0 ${N_IDX} 0
  ./eltwise_unary_reduce ${M} ${N_ADJ} ${LDI} ${REDUCE_X} ${REDUCE_X2} ${REDUCE_ROWS} ${REDUCE_OP} 1 ${N_IDX} 0
  ./eltwise_unary_reduce ${M} ${N_ADJ} 100 ${REDUCE_X} ${REDUCE_X2} ${REDUCE_ROWS} ${REDUCE_OP} 1 ${N_IDX} 0
  ./eltwise_unary_reduce ${M} ${N_ADJ} ${LDI} ${REDUCE_X} ${REDUCE_X2} ${REDUCE_ROWS} ${REDUCE_OP} 0 ${N_IDX} 0
  ./eltwise_unary_reduce ${M} ${N_ADJ} 100 ${REDUCE_X} ${REDUCE_X2} ${REDUCE_ROWS} ${REDUCE_OP} 0 ${N_IDX} 0
  ./eltwise_unary_reduce ${M} ${N_ADJ} ${LDI} ${REDUCE_X} ${REDUCE_X2} ${REDUCE_ROWS} ${REDUCE_OP} 1 ${N_IDX} 0
  ./eltwise_unary_reduce ${M} ${N_ADJ} 100 ${REDUCE_X} ${REDUCE_X2} ${REDUCE_ROWS} ${REDUCE_OP} 1 ${N_IDX} 0
  ./eltwise_unary_reduce ${M} ${N_ADJ} ${LDI} ${REDUCE_X} ${REDUCE_X2} ${REDUCE_ROWS} ${REDUCE_OP} 0 ${N_IDX} 0
  ./eltwise_unary_reduce ${M} ${N_ADJ} 100 ${REDUCE_X} ${REDUCE_X2} ${REDUCE_ROWS} ${REDUCE_OP} 0 ${N_IDX} 0
  ./eltwise_unary_reduce ${M} ${N_ADJ} ${LDI} ${REDUCE_X} ${REDUCE_X2} ${REDUCE_ROWS} ${REDUCE_OP} 1 ${N_IDX} 0
  ./eltwise_unary_reduce ${M} ${N_ADJ} 100 ${REDUCE_X} ${REDUCE_X2} ${REDUCE_ROWS} ${REDUCE_OP} 1 ${N_IDX} 0
  ./eltwise_unary_reduce ${M} ${N_ADJ} ${LDI} ${REDUCE_X} ${REDUCE_X2} ${REDUCE_ROWS} ${REDUCE_OP} 1 ${N_IDX} 0
  ./eltwise_unary_reduce ${M} ${N_ADJ} 100 ${REDUCE_X} ${REDUCE_X2} ${REDUCE_ROWS} ${REDUCE_OP} 1 ${N_IDX} 0
  ./eltwise_unary_reduce ${M} ${N_ADJ} ${LDI} ${REDUCE_X} ${REDUCE_X2} ${REDUCE_ROWS} ${REDUCE_OP} 0 ${N_IDX} 0
  ./eltwise_unary_reduce ${M} ${N_ADJ} 100 ${REDUCE_X} ${REDUCE_X2} ${REDUCE_ROWS} ${REDUCE_OP} 0 ${N_IDX} 0
  ./eltwise_unary_reduce ${M} ${N_ADJ} ${LDI} ${REDUCE_X} ${REDUCE_X2} ${REDUCE_ROWS} ${REDUCE_OP} 0 ${N_IDX} 0
  ./eltwise_unary_reduce ${M} ${N_ADJ} 100 ${REDUCE_X} ${REDUCE_X2} ${REDUCE_ROWS} ${REDUCE_OP} 0 ${N_IDX} 0
  ./eltwise_unary_reduce ${M} ${N_ADJ} ${LDI} ${REDUCE_X} ${REDUCE_X2} ${REDUCE_ROWS} ${REDUCE_OP} 1 ${N_IDX} 0
  ./eltwise_unary_reduce ${M} ${N_ADJ} 100 ${REDUCE_X} ${REDUCE_X2} ${REDUCE_ROWS} ${REDUCE_OP} 1 ${N_IDX} 0
  ./eltwise_unary_reduce ${M} ${N_ADJ} ${LDI} ${REDUCE_X} ${REDUCE_X2} ${REDUCE_ROWS} ${REDUCE_OP} 0 ${N_IDX} 0
  ./eltwise_unary_reduce ${M} ${N_ADJ} 100 ${REDUCE_X} ${REDUCE_X2} ${REDUCE_ROWS} ${REDUCE_OP} 0 ${N_IDX} 0
  ./eltwise_unary_reduce ${M} ${N_ADJ} ${LDI} ${REDUCE_X} ${REDUCE_X2} ${REDUCE_ROWS} ${REDUCE_OP} 1 ${N_IDX} 0
  ./eltwise_unary_reduce ${M} ${N_ADJ} 100 ${REDUCE_X} ${REDUCE_X2} ${REDUCE_ROWS} ${REDUCE_OP} 1 ${N_IDX} 0
  ./eltwise_unary_reduce ${M} ${N_ADJ} ${LDI} ${REDUCE_X} ${REDUCE_X2} ${REDUCE_ROWS} ${REDUCE_OP} 0 ${N_IDX} 0
  ./eltwise_unary_reduce ${M} ${N_ADJ} 100 ${REDUCE_X} ${REDUCE_X2} ${REDUCE_ROWS} ${REDUCE_OP} 0 ${N_IDX} 0
  ./eltwise_unary_reduce ${M} ${N_ADJ} ${LDI} ${REDUCE_X} ${REDUCE_X2} ${REDUCE_ROWS} ${REDUCE_OP} 1 ${N_IDX} 0
  ./eltwise_unary_reduce ${M} ${N_ADJ} 100 ${REDUCE_X} ${REDUCE_X2} ${REDUCE_ROWS} ${REDUCE_OP} 1 ${N_IDX} 0
  ./eltwise_unary_reduce ${M} ${N_ADJ} ${LDI} ${REDUCE_X} ${REDUCE_X2} ${REDUCE_ROWS} ${REDUCE_OP} 0 ${N_IDX} 0
  ./eltwise_unary_reduce ${M} ${N_ADJ} 100 ${REDUCE_X} ${REDUCE_X2} ${REDUCE_ROWS} ${REDUCE_OP} 0 ${N_IDX} 0

```

File: src/generator_mateltwise_reduce_avx_avx512.c [hunk 1]
```diff
@@ -83,6 +83,7 @@ void libxsmm_generator_reduce_cols_ncnc_avx512_microkernel( libxsmm_generated_co
     m_trips           = (bc + vlen - 1)/vlen;
     m_outer_trips     = (m_trips + m_unroll_factor - 1)/m_unroll_factor;
 
+
```
Comment (alheinecke) [inline src/generator_mateltwise_reduce_avx_avx512.c:86]: here AVX512VL support (and there BF16 is missing): libxsmm_generator_reduce_cols_ncnc_avx512_microkernel

File: src/generator_mateltwise_reduce_avx_avx512.c [hunk 2]
```diff
@@ -415,16 +416,28 @@ void libxsmm_generator_reduce_cols_avx512_microkernel( libxsmm_generated_code*
                                    (i_mateltwise_desc->param == LIBXSMM_MELTW_TYPE_UNARY_REDUCE_X2_OP_ADD)) ? 1 : 0;
   unsigned int flag_reduce_op_max = (i_mateltwise_desc->param == LIBXSMM_MELTW_TYPE_UNARY_REDUCE_X_OP_MAX) ? 1 : 0;
 
-  if (LIBXSMM_DATATYPE_BF16 == LIBXSMM_GETENUM_INP( i_mateltwise_desc->datatype )) {
-    vname_in = 'y';
-    vmove_instruction_in = LIBXSMM_X86_INSTR_VMOVDQU16;
-  }
+  if ( (io_generated_code->arch > LIBXSMM_X86_AVX512_VL256) && (io_generated_code->arch < LIBXSMM_X86_AVX512) ) {
```
Comment (alheinecke) [inline src/generator_mateltwise_reduce_avx_avx512.c:419]: That should be >= AVX512_VL256

File: src/generator_mateltwise_reduce_avx_avx512.c [hunk 3]
```diff
@@ -1064,6 +1093,9 @@ void libxsmm_generator_reduce_rows_avx512_microkernel( libxsmm_generated_code*
   i_gp_reg_mapping->gp_reg_reduced_elts_squared = LIBXSMM_X86_GP_REG_R10;
   i_gp_reg_mapping->gp_reg_m_loop               = LIBXSMM_X86_GP_REG_R11;
   i_gp_reg_mapping->gp_reg_n_loop               = LIBXSMM_X86_GP_REG_RAX;
+  int bf16_accum = LIBXSMM_X86_GP_REG_RCX;
+  libxsmm_x86_instruction_alu_imm( io_generated_code, i_micro_kernel_config->alu_sub_instruction, LIBXSMM_X86_GP_REG_RSP, 128 );
```
Comment (alheinecke) [inline src/generator_mateltwise_reduce_avx_avx512.c:1097]: That seems to be scratch allocation. Why is that not handled in the scratch allocation framework.

File: src/generator_mateltwise_reduce_avx_avx512.c [hunk 4]
```diff
@@ -1118,12 +1150,23 @@ void libxsmm_generator_reduce_rows_avx512_microkernel( libxsmm_generated_code*
       libxsmm_generator_load_vreg_minus_infinity(io_generated_code, 'y', LIBXSMM_X86_GP_REG_RAX, aux_vreg);
     }
 
-    libxsmm_generator_mateltwise_initialize_avx_mask(io_generated_code, mask_out, 1);
-    /* Calculate input mask in case we see m_masking */
-    if (use_m_masking == 1) {
-      mask_reg = available_vregs-1;
-      libxsmm_generator_mateltwise_initialize_avx_mask(io_generated_code, mask_reg, m % vlen);
-      available_vregs--;
+    if (io_generated_code->arch <= LIBXSMM_X86_AVX2) {
```
Comment (alheinecke) [inline src/generator_mateltwise_reduce_avx_avx512.c:1153]: This if doesn't make sense, we are already only AVX2 or lower due to the if in line 1133

File: src/generator_mateltwise_reduce_avx_avx512.c [hunk 5]
```diff
@@ -1149,13 +1192,23 @@ void libxsmm_generator_reduce_rows_avx512_microkernel( libxsmm_generated_code*
...
         /* If we have remainder, then we want to blend in -INF for the zero'ed out entries */
         if ((use_m_masking == 1) && (im == (m_trips-1))) {
-          libxsmm_x86_instruction_vec_compute_3reg_mask_sae_imm8(io_generated_code,
-                    LIBXSMM_X86_INSTR_VBLENDVPS,
-                    'y',
-                    reg_sum,
-                    aux_vreg,
-                    reg_sum,
-                    0, 0, 0, (mask_reg) << 4 );
+          if (io_generated_code->arch <= LIBXSMM_X86_AVX2) {
```
Comment (alheinecke) [inline src/generator_mateltwise_reduce_avx_avx512.c:1195]: same issue

File: src/generator_mateltwise_reduce_avx_avx512.c [hunk 6]
```diff
@@ -1179,13 +1232,23 @@ void libxsmm_generator_reduce_rows_avx512_microkernel( libxsmm_generated_code*
...
       /* If we have remainder, then we want to blend in -INF for the zero'ed out entries */
       if ( (flag_reduce_op_max > 0) && (im == m_trips-1) && (use_m_masking > 0)) {
-        libxsmm_x86_instruction_vec_compute_3reg_mask_sae_imm8(io_generated_code,
-                  LIBXSMM_X86_INSTR_VBLENDVPS,
-                  'y',
-                  cur_vreg,
-                  aux_vreg,
-                  cur_vreg,
-                  0, 0, 0, (mask_reg) << 4 );
+          if (io_generated_code->arch <= LIBXSMM_X86_AVX2) {
```
Comment (alheinecke) [inline src/generator_mateltwise_reduce_avx_avx512.c:1235]: same issue

File: src/generator_mateltwise_reduce_avx_avx512.c [hunk 7]
```diff
@@ -1254,40 +1317,46 @@ void libxsmm_generator_reduce_rows_avx512_microkernel( libxsmm_generated_code*
...
   /* We fully unroll in M dimension, calculate mask if there is remainder */
-  m                 = i_mateltwise_desc->m;
-  n                 = i_mateltwise_desc->n;
-  use_m_masking     = ( m % 16 == 0 ) ? 0 : 1;
-  use_n_masking     = ( n % 16 == 0 ) ? 0 : 1;
-  m_trips           = (m + 15) / 16;
-  n_trips           = (n + 15) / 16;
-  n_full_trips      = ( n % 16 == 0 ) ? n_trips : n_trips-1;
-
+    m                 = i_mateltwise_desc->m;
```
Comment (alheinecke) [inline src/generator_mateltwise_reduce_avx_avx512.c:1320]: wrong indention

File: src/generator_mateltwise_reduce_avx_avx512.c [hunk 8]
```diff
@@ -1254,40 +1317,46 @@ void libxsmm_generator_reduce_rows_avx512_microkernel( libxsmm_generated_code*
...
+    n                 = i_mateltwise_desc->n;
+    use_m_masking     = ( m % vlen == 0 ) ? 0 : 1;
+    use_n_masking     = ( n % vlen == 0 ) ? 0 : 1;
+    m_trips           = ( m+vlen-1 )/ vlen;
+    n_trips           = ( n+vlen-1 )/ vlen;
+    n_full_trips      = ( n % vlen == 0 ) ? n_trips : n_trips-1;
+    // n_cols_load = vlen;
```
Comment (alheinecke) [inline src/generator_mateltwise_reduce_avx_avx512.c:1327]: C++ comments are not allowed.

File: src/generator_mateltwise_reduce_avx_avx512.c [hunk 9]
```diff
@@ -1723,145 +1837,146 @@ void libxsmm_generator_reduce_rows_avx512_microkernel( libxsmm_generated_code*
 
       /* 2nd stage */
       /* zmm0/zmm8; 4444 4444 0000 0000 / cccc cccc 8888 8888  -> zmm0: cccc 8888 4444 0000 */
-      if ( compute_plain_vals_reduce > 0 ) {
-        libxsmm_x86_instruction_vec_compute_3reg_imm8( io_generated_code,
-                                                 LIBXSMM_X86_INSTR_VSHUFF64X2,
-                                                 i_micro_kernel_config->vector_name,
-                                                 8, 0, 16, 0x88 );
+      if(io_generated_code->arch >LIBXSMM_X86_AVX512){
```
Comment (alheinecke) [inline src/generator_mateltwise_reduce_avx_avx512.c:1840]: That should be >=

File: src/generator_mateltwise_reduce_avx_avx512.c [hunk 10]
```diff
@@ -3050,16 +3235,61 @@ void libxsmm_generator_reduce_rows_avx512_microkernel( libxsmm_generated_code*
...
     }
   }
 
   if (LIBXSMM_DATATYPE_BF16 == LIBXSMM_GETENUM_OUT( i_mateltwise_desc->datatype )) {
-    if (io_generated_code->arch < LIBXSMM_X86_AVX512_CPX) {
+    if ((io_generated_code->arch < LIBXSMM_X86_AVX512_CPX) && (io_generated_code->arch != LIBXSMM_X86_AVX512_VL256_CPX)) {
       libxsmm_generator_vcvtneps2bf16_avx512_clean_stack( io_generated_code, LIBXSMM_X86_GP_REG_RAX );
     }
   }
+  libxsmm_x86_instruction_alu_imm( io_generated_code, i_micro_kernel_config->alu_add_instruction, LIBXSMM_X86_GP_REG_RSP, 128 );
```
Comment (alheinecke) [inline src/generator_mateltwise_reduce_avx_avx512.c:3292]: see the comment on the scratch above

File: src/generator_mateltwise_reduce_avx_avx512.c [hunk 11]
```diff
@@ -1628,10 +1732,15 @@ void libxsmm_generator_reduce_rows_avx512_microkernel( libxsmm_generated_code*
                                                      i_micro_kernel_config->vector_name,
                                                      7, 3, 16, 7, 0 );
 
+/*         libxsmm_x86_instruction_vec_compute_3reg_imm8( io_generated_code,
```
Comment (alheinecke) [inline src/generator_mateltwise_reduce_avx_avx512.c:1735]: can that be deleted?

File: src/generator_mateltwise_reduce_avx_avx512.c [hunk 12]
```diff
@@ -1651,10 +1760,15 @@ void libxsmm_generator_reduce_rows_avx512_microkernel( libxsmm_generated_code*
                                                      i_micro_kernel_config->vector_name,
                                                      19, 18, 20, 7, 0 );
 
+/*         libxsmm_x86_instruction_vec_compute_3reg_imm8( io_generated_code,
```
Comment (alheinecke) [inline src/generator_mateltwise_reduce_avx_avx512.c:1763]: can that be deleted?

**Review by alheinecke — CHANGES_REQUESTED** [review]
multiple issues still exist

File: src/generator_mateltwise_reduce_avx_avx512.c (hunk 1 above)
Comment (deeptiag1) [reply src/generator_mateltwise_reduce_avx_avx512.c:86]: Yes this hasn't been promoted yet

File: src/generator_mateltwise_reduce_avx_avx512.c (hunk 2 above)
Comment (deeptiag1) [reply src/generator_mateltwise_reduce_avx_avx512.c:419]: Corrected in commit

File: src/generator_mateltwise_reduce_avx_avx512.c (hunk 3 above)
Comment (deeptiag1) [reply src/generator_mateltwise_reduce_avx_avx512.c:1097]: @egeor Shall we switch this to the main function?

**Comment by deeptiag1** [comment]
Seg fault is fixed. Cleaned the code.
Script should pass now.

File: src/generator_mateltwise_reduce_avx_avx512.c (hunk 2 above)
Comment (alheinecke) [reply src/generator_mateltwise_reduce_avx_avx512.c:419]: Still seems to be incorrect... not sure where it's corrected but in the PR line 419 is incorrect.

File: src/generator_mateltwise_reduce_avx_avx512.c (hunk 3 above)
Comment (egeor) [reply src/generator_mateltwise_reduce_avx_avx512.c:1097]: Assuming the code works, you can use for the stack management the following framework we have for stack management:
https://github.com/libxsmm/libxsmm/blob/master/src/generator_mateltwise_sse_avx_avx512.c#L112
https://github.com/libxsmm/libxsmm/blob/master/src/generator_mateltwise_sse_avx_avx512.c#L183
https://github.com/libxsmm/libxsmm/blob/master/src/generator_mateltwise_sse_avx_avx512.c#L99
https://github.com/libxsmm/libxsmm/blob/master/src/generator_mateltwise_sse_avx_avx512.c#L86

You'd need basically to allocate scratch by adjusting lines (to allocate 128 bytes when executing reduce rows kernel):
https://github.com/libxsmm/libxsmm/blob/master/src/generator_mateltwise_sse_avx_avx512.c#L123
https://github.com/libxsmm/libxsmm/blob/master/src/generator_mateltwise_sse_avx_avx512.c#L156

and then use LIBXSMM_MELTW_STACK_VAR_SCRATCH_PTR stack var name in the libxsmm_generator_meltw_getval_stack_var/libxsmm_generator_meltw_setval_stack_var calls.

**Review by alheinecke — CHANGES_REQUESTED** [review]
Still one issue (line 419) doesn't seem to be corrected. 
What's the ETA on the missing routine?
What's the ETA on the stack allocation?

**Comment by alheinecke** [comment]
Why is the segfault going away I don't see any fixes which should make a segfault go away also it's not clear to me how 
` ./samples/eltwise/eltwise_unary_reduce 56 79 56 56 1 0 1 0 0 0 0`
was generated by the script.

File: src/generator_mateltwise_reduce_avx_avx512.c (hunk 3 above)
Comment (deeptiag1) [reply src/generator_mateltwise_reduce_avx_avx512.c]: Thanks @egeor

**Comment by deeptiag1** [comment]
To clear the confusion on Segfault:
In my test output I saw only this 56 79 56 56 so this may not be command produced by the test script. 
I was getting Segfault for ./samples/eltwise/eltwise_unary_reduce 31 8 31 0 1 1 0 0 0 0. This was happening bcoz. of code error at https://github.com/deeptiag1/libxsmm/blob/2999b288e3e79a38c85a49795cdbc23bcf999f30/src/generator_mateltwise_reduce_avx_avx512.c#L3182

File: src/generator_gemm_common.c [hunk 13]
```diff
@@ -35,10 +36,16 @@ void libxsmm_generator_gemm_vnni_store_C_from_scratch( libxsmm_generated_code*
   libxsmm_generator_gemm_getval_stack_var( io_generated_code, i_micro_kernel_config, LIBXSMM_GEMM_STACK_VAR_GEMM_SCRATCH_PTR, l_gp_reg_in );
   libxsmm_x86_instruction_alu_imm_i64( io_generated_code, LIBXSMM_X86_INSTR_MOVQ, i_gp_reg_mapping->gp_reg_help_1, 32*64 );
   libxsmm_x86_instruction_alu_reg( io_generated_code, i_micro_kernel_config->alu_add_instruction, i_gp_reg_mapping->gp_reg_help_1, l_gp_reg_in);
-  libxsmm_generator_transform_norm_to_vnni_16bit_avx512_microkernel( io_generated_code, io_loop_label_tracker,
```
Comment (alheinecke) [inline src/generator_gemm_common.c:38]: Why do we need to touch the GEMM code generator? I don't understand this.

**Review by alheinecke — COMMENTED** [review]
not sure about the latest changes to GEMM generator

**Pull request merged**